src/webui/www/public/scripts/login.js

```diff
@@ -30,32 +30,32 @@
 
 const submitLoginForm = (event) => {
     event.preventDefault();
-    const errorMsgElement = document.getElementById("error_msg");
 
-    const xhr = new XMLHttpRequest();
-    xhr.open("POST", "api/v2/auth/login", true);
-    xhr.setRequestHeader("Content-type", "application/x-www-form-urlencoded; charset=UTF-8");
-    xhr.addEventListener("readystatechange", () => {
-        if (xhr.readyState === 4) { // DONE state
-            if ((xhr.status === 200) && (xhr.responseText === "Ok."))
-                location.replace(location);
-            else
-                errorMsgElement.textContent = "QBT_TR(Invalid Username or Password.)QBT_TR[CONTEXT=Login]";
-        }
-    });
-    xhr.addEventListener("error", () => {
-        errorMsgElement.textContent = (xhr.responseText !== "")
-            ? xhr.responseText
-            : "QBT_TR(Unable to log in, server is probably unreachable.)QBT_TR[CONTEXT=Login]";
-    });
+    const errorMsgElement = document.getElementById("error_msg");
+    errorMsgElement.textContent = ""; // clear previous error
 
     const usernameElement = document.getElementById("username");
     const passwordElement = document.getElementById("password");
-    const queryString = "username=" + encodeURIComponent(usernameElement.value) + "&password=" + encodeURIComponent(passwordElement.value);
-    xhr.send(queryString);
+    const queryString = `username=${encodeURIComponent(usernameElement.value)}&password=${encodeURIComponent(passwordElement.value)}`;
+    passwordElement.value = ""; // clear previous value
 
-    // clear the field
-    passwordElement.value = "";
+    fetch("api/v2/auth/login", {
+            method: "POST",
+            headers: {
+                "Content-type": "application/x-www-form-urlencoded; charset=UTF-8"
+            },
+            body: queryString
+        })
+        .then(async (response) => {
+                const responseText = await response.text();
+                if (response.ok && (responseText === "Ok."))
+                    location.replace(location); // redirect
+                else
+                    errorMsgElement.textContent = `QBT_TR(Invalid Username or Password.)QBT_TR[CONTEXT=Login]\nQBT_TR(Server response:)QBT_TR[CONTEXT=Login] ${responseText}`;
+            },
+            (error) => {
+                errorMsgElement.textContent = `QBT_TR(Unable to log in, server is probably unreachable.)QBT_TR[CONTEXT=Login]\n${error}`;
+            });
```

Comment:
It would be preferable to use `await` here to avoid nesting.

```suggestion
        try {
            const response = await fetch("api/v2/auth/login", {
                method: "POST",
                headers: {
                    "Content-type": "application/x-www-form-urlencoded; charset=UTF-8"
                },
                body: queryString
            });

            const responseText = await response.text();
            if (response.ok && (responseText === "Ok."))
                location.replace(location); // redirect
            else
                errorMsgElement.textContent = `QBT_TR(Invalid Username or Password.)QBT_TR[CONTEXT=Login]\nQBT_TR(Server response:)QBT_TR[CONTEXT=Login] ${responseText}`;
        }
        catch(error) {
            errorMsgElement.textContent = `QBT_TR(Unable to log in, server is probably unreachable.)QBT_TR[CONTEXT=Login]\n${error}`;
        }
```

Reply:
I imagine he did it this way to avoid declaring the parent function as `async` and having to `await` on every callsite of the parent function.

Reply:
That's a good point. I just checked and `submitLoginForm` is only passed to `addEventListener`, so there shouldn't be any impact.

Reply:
>I imagine he did it this way to avoid declaring the parent function as async and having to await on every callsite of the parent function.

Yes. Also, function using `async` means it will return a `Promise` and it somewhat changes the function intention/signature. IMO async code should be contained in limited scope to avoid `async` keyword pollution.

Another minor one is, The `try catch` doesn't good to me. It should only surround `fetch()` but it cannot due to `response` has limited scope.

Using `.then()` doesn't look too bad. It clearly shows the success case and the fail case. It is similar to `if else`. 
So, if there is no strong objection I would like to keep it as is.

Interesting read: https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/

Reply:
>The function already runs asynchronously due to the use of .then, so it's equivalent to calling an async function without awaiting it.

True but it doesn't mean that the function needs to be marked `async`. The async nature can be an internal/local implementation.